rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala

```diff
@@ -25,34 +26,30 @@ object implicits {
       }
     }
 
-  def matchListPar(init: Cost): StorageMatch[
-    BindPattern,
-    InterpreterError,
-    ListParWithRandom,
-    ListParWithRandomAndPhlos
-  ] =
-    new StorageMatch[
-      BindPattern,
-      InterpreterError,
-      ListParWithRandom,
-      ListParWithRandomAndPhlos
-    ] {
+  def matchListPar[F[_]: Sync](
+      init: Cost
+  ): StorageMatch[F, BindPattern, ListParWithRandom, ListParWithRandomAndPhlos] =
+    new StorageMatch[F, BindPattern, ListParWithRandom, ListParWithRandomAndPhlos] {
 
       private def calcUsed(init: Cost, left: Cost): Cost = init - left
 
       def get(
           pattern: BindPattern,
           data: ListParWithRandom
-      ): Either[InterpreterError, Option[ListParWithRandomAndPhlos]] =
-        SpatialMatcher
-          .foldMatch[NonDetFreeMapWithCost, Par, Par](
-            data.pars,
-            pattern.patterns,
-            pattern.remainder
-          )
-          .runFirstWithCost(init)
-          .map {
-            case (left, resultMatch) =>
+      ): F[Option[ListParWithRandomAndPhlos]] =
+        Sync[F]
+          .delay {
+            SpatialMatcher
+              .foldMatch[NonDetFreeMapWithCost, Par, Par](
```

Comment:
we should require `:_cost` or `:CostAccounging` in here, and `s/NonDetFreeMapWithCost/MatcherMonadT[F, ?]`. In the same or immediately next step we should also no longer return `F[Option[ListParWithRandomAndPhlos]]` and just return `F[Option[ListParWithRandom]]`.

Reply:
I agree. That's the PR after this one :)